pgrollup/parser_pg.py

```diff
@@ -0,0 +1,290 @@
+import sys
+from lark import Lark, Transformer
+
+def parse_create(sql):
+    '''
+    Example expected output:
+    [{'joininfos': '[{"table_name": "testjoin1", "table_alias": "t1", "condition": "", "join_type": "FROM"}, 
+    {"table_name": "testjoin2", "table_alias": "t2", "condition": "using (id)", "join_type": "INNER JOIN"}]',
+    'rollup_name': 'testjoin_rollup1',
+    'groups': [['t1.name', '"t1.name"']],
+    'columns': [['sum(t1.num)', 'sum_num'], ['sum(t2.foo)', 'sum_foo']],
+    'where_clause': None, 'having_clause': None}]
+    '''
+    dict = parse(sql)
+    infos = []                                              
+    infos.append({
+        'joininfos' :_getjoins(dict[0]['stmt']['fromClause'][0]),
+#       'groups' : [[x['fields'][0],str(x['fields'][0])] for x in dict[0]['stmt']['groupClause']],
+        'groups' : _getgroups(dict),
+        'columns': [[expr_to_str(x['val']),_getalias(x)] for x in dict[0]['stmt']['targetList']],
+        'where_clause': _getwhere(dict),
+        'having_clause': _gethaving(dict),
+        })
+    return infos
+
+
+#Lark automagically creates a tree that represents the parsed text according to the grammar
+grammar = r"""
+    ?value: dict
+          | list
+          | ESCAPED_STRING     ->string
+          | NAME               -> name   
+          | SIGNED_NUMBER      -> number
+          | "true"             -> true
+          | "false"            -> false
+          | "<>"             -> null
+    
+    list: "(" [value*] ")"
+    dict: "{" value [pair*] "}"
+    pair : ":"NAME value
+
+    NAME: /[a-zA-Z_.0-9]+/
+    
+
+    %import common.ESCAPED_STRING
+    %import common.SIGNED_NUMBER
+    %import common.WS
+    %ignore WS
+    """
+
+class Transformer(Transformer):
+    '''
+    A class with methods corresponding to branch names. 
+    For each branch, the appropriate method will be called 
+    with the children of the branch as its argument, and 
+    its return value will replace the branch in the tree.
+    '''
+    def string(self, s):
+        '''
+        Example input:
+        [Tree('string', [Token('ESCAPED_STRING', '"t1"')]),Tree('string', [Token('ESCAPED_STRING', '"num"')])]
+        Example output:
+        ['t1','num']
+       '''
+        (s,) = s
+        return s[1:-1]
+    def number(self, n):
+        '''
+        Example input:
+        Tree('number', [Token('SIGNED_NUMBER', '40')])
+        Example output:
+        40
+        '''
+        (n,) = n
+        return int(n)                        
+    def pair(self,key_value):
+        k, v = key_value
+        return str(k),v
+                       
+    list = list   
+#   pair = tuple
+    dict = dict
+
+    #NAME = lambda self, n: n
+    name  = lambda self,n: ('TYPE',str(n[0]))
+    null = lambda self, _: None
+    true = lambda self, _: True
+    false = lambda self, _: False
+
+parser = Lark(grammar, start='value', lexer='standard')
+
+import pprint
```

Comment:
do not include debugging imports in production code;

they needlessly slow down the loading of the code;

for one file that's not noticeable, but over the course of a large project, these slow downs add up considerably